spinoso-string/src/enc/utf8/mod.rs

```diff
@@ -287,6 +287,17 @@ impl Utf8String {
     pub fn get_char_slice(&self, range: Range<usize>) -> Option<&'_ [u8]> {
         // TODO: use fast path rejection from `get_char` here too
         let Range { start, end } = range;
+
+        if start > self.len() || start > self.char_len() {
+            return None;
+        }
```

Comment:
in the old code, this check was nested within a `if end < start` condition so we can avoid a UTF-8 decode unless we know we're short circuiting.

```suggestion
        // Fast path the lookup if the end of the range is before the start.
        if end < start {
            // Yes, these types of ranges are allowed and they return `""`,
            // but only if `start` is within the string.
            //
            // Attempt to short-circuit with a cheap length retrieval.
            if start > self.len() || start > self.char_len() {
                return None;
            }
            return Some(&[]);
        }
```